protocols/virtualfund/virtualfund.go

```diff
@@ -344,39 +340,56 @@ func (o Objective) Channels() []*channel.Channel {
 //  Private methods on the VirtualFundObjective //
 //////////////////////////////////////////////////
 
+// insertGuaranteeInfo mutates the reciever Connection struct.
+func (connection *Connection) insertGuaranteeInfo(a0 types.Funds, b0 types.Funds, vId types.Destination, left types.Destination, right types.Destination) error {
+
+	connection.GuaranteeInfo = GuaranteeInfo{
+		Left:                 left,
+		Right:                right,
+		LeftAmount:           a0,
+		RightAmount:          b0,
+		GuaranteeDestination: vId,
+	}
+
+	// Check that the guarantee metadata can be encoded. This allows us to avoid clunky error-return-chains for getExpectedGuarantees
+	metadata := outcome.GuaranteeMetadata{
+		Left:  connection.GuaranteeInfo.Left,
+		Right: connection.GuaranteeInfo.Right,
+	}
+	_, err := metadata.Encode()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
 // insertExpectedGuaranteesForLedgerChannel mutates the reciever Connection struct.
-func (connection *Connection) insertExpectedGuarantees(a0 types.Funds, b0 types.Funds, vId types.Destination, left types.Destination, right types.Destination) error {
+func (connection *Connection) getExpectedGuarantees() map[types.Address]outcome.Allocation {
 	expectedGuaranteesForLedgerChannel := make(map[types.Address]outcome.Allocation)
 	metadata := outcome.GuaranteeMetadata{
-		Left:  left,
-		Right: right,
+		Left:  connection.GuaranteeInfo.Left,
+		Right: connection.GuaranteeInfo.Right,
 	}
 	encodedGuarantee, err := metadata.Encode()
+	// This error is unexpected. insertGuaranteeInfo checks that the guarantee metadata can be encoded.
+	// If this panic is triggered, GuranteeInfo has been modified after creation
 	if err != nil {
-		return err
+		panic(err)
```

Comment:
I suppose we can't be 100% sure that this is the reason, if we get a panic. What if the metadata cannot be encoded, but the first time we try, the error is swallowed somewhere in the call stack? 

The check performed in `insertGuaranteeInfo` seems quite far away from this panic, which makes me a bit uneasy about using that check as justification for the panic.

That said, I think this approach is fine. We don't really expect any genuine errors encoding guarantee metadata, so panicking actually seems like the right thing to do in _any case_, because it probably indicates a programming error / bug.

Reply:
I agree with your thoughts. One interesting question is why metadata encoding is here at all. `getExpectedGuarantees` is only used to check whether a channel affords the guarantees. It seems that encoding is not necessary for that operation. However, the `Affords` api accepts an `Allocation` with encoded metadata: https://github.com/statechannels/go-nitro/blob/81cf404220048c83e6fe3e1b153cc2130340ae05/channel/state/outcome/exit.go#L177-L179

My larger takeaways are the following:
1. If a low-level operation errors, error handling becomes cumbersome. Every function that uses the operation must now return an error.
2. One strategy is to ensure low-level operations do not error.

In this particular case, encoding of metadata can be isolated to logic that need to encode metadata. Encoding must take place for data authentication (signatures on state) and submission to chain.